Removing all nonmigratable instances of $scope and $rootScope from skill-editor module

srijanreddy98 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #N/A.
2. This PR does the following: [Explain here what your PR does and why]
This pr is an extension from the discussion in #9048.

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## PR Pointers

- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays
- Never force push. If you do, your PR will be closed.

srijanreddy98 committed:
Skill Init Event

bansalnitish reviewed (commented):
Well these changes look nice to me.

I've one ques: Can't we migrate the service `skill-editor-state.service.ts` to Angular completely? I understand it does depend on some other services which have not been migrated but all such non migrated services were blocked because of the $rootScope / $scope issue. Can we follow the dependency order and migrate the services as a whole? 

I see one concern about my question. It looks like the all such services might not be in our hand. Say for example a service A has some event and is using $rootScope. There is some service B that too triggers/listens to the same event and is dependent on some other service C which is not migrated, then we can migrate service A but we need to change the $scope(s) in service B to use the subjects itself without migrating it (if will plan not to migrate all services in a bunch which is the ideal way), right? What are your thoughts?

bansalnitish commented:
Also adding @seanlip and @vojtechjelinek to take a pass over these changes since they were involved in the discussion related to migration of $rootScope.

seanlip commented on core/templates/pages/skill-editor-page/navbar/skill-editor-navbar.directive.ts:
```diff
@@ -145,7 +145,9 @@ angular.module('oppia').directive('skillEditorNavbar', [
             $scope.skill = SkillEditorStateService.getSkill();
             $scope.validationIssues = [];
-            $scope.$on(EVENT_SKILL_INITIALIZED, _validateSkill);
+            SkillEditorStateService.getEventSkillInitializedSubject().subscribe(
```
While using subjects might make sense in general, I thought the point of this PR was to remove the need for the event altogether, by having the skill editor service store the skill info and the controller/directive retrieve a reference to it? The pattern you've provided here seems reasonable in the general case (though maybe add me as reviewer to other PRs which use it so that we can make sure we're going in the right direction), but for this specific case I think that using the Subject infrastructure (or events at all) might be more convoluted than necessary, given what the code is doing.

srijanreddy98 replied:
@seanlip, Got it! I will make the changes.
I have a couple more questions:
Q1>If this were angular 8, I would be in favor of removing this `$scope.validationIssues = [];` and `$scope.skill`.
Instead, I would add an `@Input()` to skill-navbar-directive and questions-tab-directive, and keep the skill object in the controller itself. That way no event will be required. Any change to the object will be passed on.

Is there is an equivalent of this `@Input()` in angularjs? 


> by having the skill editor service store the skill info and the controller/directive retrieve a reference to it?

Q2> How do we plan to notify the directive of changes in the object? (Using watch could complicate the migration process).
My idea would be to use subjects but what do you suggest I use.

seanlip replied:
Q1) I don't know how @Input() works, sorry. This is something you might need to ask @bansalnitish. 

Q2) Ideally you set it up so that you don't have to notify the directive. You can do this by having the service return a reference that is automatically updated when the value is changed. If I remember correctly, one way we used to do this was with a `var data = {currentSkill: ...}` dict and return data.currentSkill. You'll need to check that, if data.currentSkill is subsequently changed, the resulting value is too. (That said ... we might need to figure out how to update the derived data too -- maybe we can link it to the data dict as well and provide similar getters in the API.)

Can you explain why a watch would complicate the migration process?

Also, one thought I had while writing this is that an issue we have is that we don't know what this would, ideally, look like in Angular 8. So I guess I have a question for you, which is: if you were implementing this functionality from scratch, so that it behaves the same way as the code here (but not necessarily following the same coding patterns), how would you do it? If we know the desired end result then we can chart a pathway from where we are now to that end result.

srijanreddy98 replied:
> Can you explain why a watch would complicate the migration process?

The $watch has no equivalent in Angular, we will have to look for workarounds/ deal with it case by case. It would be better not to introduce more of watch.
> Also, one thought I had while writing this is that an issue we have is that we don't know what this would, ideally, look like in Angular 8. So I guess I have a question for you, which is: if you were implementing this functionality from scratch, so that it behaves the same way as the code here (but not necessarily following the same coding patterns), how would you do it? If we know the desired end result then we can chart a pathway from where we are now to that end result.

In this particular case, we would be using what I asked about in the first question in my previous comment. @Input, i.e. property binding... There will be a stackblitz link, **in the end**, showcasing what I think would be ideal.
In the stackblitz app there are two (important) components, app.component, and the-nav-bar.component

In the-nav-bar.component.ts, I have put this
```
  @Input() skill = {
    name: 'Placeholder Skill Name',
    value: 'Placeholder Skill Value'
  };
```
So what is will basically do is that it will open a property in the-nav-bar selector so you can pass in data to that component from the parent component (the component where the tag <the-nav-bar> is used in HTML). 

In order to pass data, you will have to bind the property by using square brackets []. I have done that in app.component.html,
```
<the-nav-bar [skill] = "skillObj"></the-nav-bar>
```
In app.component.ts I have made a  skillObj:
```
skillObj = {
    name: 'Skill Name',
    value: 'Skill Value'
  };
```
and in the same file I have put two ways if changing the object:
One through setTimeout and the other through a button,
```
ngOnInit() {
    setTimeout(
      () => {
        this.skillObj.name = 'Skill name Changed';
      }, 4000
    );
  }
  changeSkillValue() {
    this.skillObj.value = 'skill value changed';
  }
```

So within 4 seconds of the app loading, it will change the skill name to `skill name changed` and if you click a button saying `Click here to change skill value` it will change the skill value to `skill value changed`.

**Please note that I haven't explicitly made any attempt to reach the-nav-bar.component to inform it of changes happening in the skillObj, it happens automatically because of angular change detection, which is highly optimized**.

StackBlitz link: https://stackblitz.com/edit/angular-g5e6pl
Angular guide link: https://angular.io/guide/component-interaction

srijanreddy98 replied:
+cc @bansalnitish ptal at my comment and let me know if you agree or if you would go for a different way.

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into migrate-skill-init-event

kevinlee12 commented:
@vojtechjelinek @bansalnitish ptal!

srijanreddy98 committed:
Changes

srijanreddy98 committed:
Linting issues

srijanreddy98 commented:
@seanlip, @bansalnitish  since the `EVENT_SKILL_INITIALIZED` exists in the skill-editor-navbar.directive.ts, using `bindToController` is not possible at this moment because it is used in skill-editor-mainpage.html. The controller in skill-editor-mainpage.html is from the base-content.directive.ts and we can use module-specific variables in base-content.directive.ts. So I think we will have to use subjects in this case.

seanlip commented:
OK, do what you need to do, and add detailed explanatory comments stating (a) why it's done this way, and (b) what you want to change it to after XXX is completed (specify what XXX is). This is so that this can be handled in a better way later.

srijanreddy98 commented:
@seanlip I am thinking of raising an issue to keep track of this and add a TODO in the code base wherever subjects were not appropriate. (in the context of ($scope and $rootScope).($on\$broadcast) migration.

seanlip commented:
As long as things are fully documented so that a newcomer to the project
can grok what's going on easily and there's no confusion, I'm fine with any
documentation approach you take. Just make sure that the plan is completely
clear. Thanks!

bansalnitish commented:
I'm fine with using subjects but as you said we have to use subjects at this moment and later we would change it to bindings -- could we please properly document with a clear example as to why we are using subjects now? When would they be changed to bindings and steps on changing them from subjects to bindings? 

Ping me to review it when you're done. I think compiling a one-two pager doc would be much useful but I'm fine either way (directly opening an issue). You can link that doc to an issue then.

srijanreddy98 committed:
Remove all instances of  and  except

srijanreddy98 committed:
Resolve conflicts

bansalnitish commented on core/templates/pages/skill-editor-page/editor-tab/skill-concept-card-editor/skill-concept-card-editor.directive.ts:
```diff
@@ -210,6 +216,10 @@ angular.module('oppia').directive('skillConceptCardEditor', [
             };
           };
+
+          $scope.$on('$destroy', function() {
```
Please add corresponding equivalent of this somewhere so that we do not forget this (while we open migration of directives to new contributors we can such equivalents there then)

seanlip replied:
Hi @bansalnitish -- just to check, what do you mean by this comment? (What is "this", and where is the "somewhere" you are referring to?)

srijanreddy98 replied:
@seanlip, Maybe @bansalnitish meant what is the equivalent of `$scope.$on('$destroy'` in angular. This entire `$scope.$on('$destroy'` can be replaced with the `ngOnDestroy` lifecycle hook and maybe he wants it to be added to the angular migration doc so that new contributors don't get stuck or confused.

bansalnitish reviewed (approved):
LGTM!

kevintab95 commented:
@srijanreddy98 please mind the Travis failures!

srijanreddy98 committed:
Remove unused imports

srijanreddy98 committed:
Change subject names

srijanreddy98 committed:
Change subject names

kevintab95 commented on core/templates/pages/skill-editor-page/questions-tab/skill-questions-tab.directive.ts:
```diff
@@ -75,9 +75,23 @@ angular.module('oppia').directive('questionsTab', [
             _init();
-            $scope.$on(EVENT_SKILL_INITIALIZED, _init);
-            $scope.$on(EVENT_SKILL_REINITIALIZED, _init);
+            ctrl.parentSubscription.add(
```
```suggestion
            ctrl.parentSubscription.add(
              SkillEditorStateService.getSkillReinitializedSubject().subscribe(
                () => _init()));
```

srijanreddy98 replied:
Done!

srijanreddy98 replied:
And also, I kept the last `);` in a separate line. to shows where the `ctrl.parentSubscription.add(` is ending.

kevintab95 commented on core/templates/pages/skill-editor-page/questions-tab/skill-questions-tab.directive.ts:
```diff
@@ -75,9 +75,23 @@ angular.module('oppia').directive('questionsTab', [
+                )
+            );
+            ctrl.parentSubscription.add(
+              SkillEditorStateService.getSkillReinitializedSubject()
```
Always break after `(` . Please follow the same pattern suggested above? Thanks!

srijanreddy98 replied:
Done!

kevintab95 commented on core/templates/pages/skill-editor-page/questions-tab/skill-questions-tab.directive.ts:
```diff
@@ -34,6 +34,7 @@ require(
 require('services/alerts.service.ts');
 require('services/questions-list.service.ts');
 require('services/contextual/url.service.ts');
+import { Subscription } from 'rxjs';
```
Add newline above this.

srijanreddy98 replied:
Done!

kevintab95 commented on core/templates/pages/skill-editor-page/editor-tab/skill-rubrics-editor/skill-rubrics-editor.directive.ts:
```diff
@@ -35,22 +37,30 @@ angular.module('oppia').directive('skillRubricsEditor', [
-              $scope.rubrics = $scope.skill.getRubrics();
-            });
+            ctrl.parentSubscription.add(
+              SkillEditorStateService.getSkillReinitializedSubject()
```
```suggestion
              SkillEditorStateService.getSkillReinitializedSubject().subscribe(
                () => $scope.rubrics = $scope.skill.getRubrics()));
```

kevintab95 commented on core/templates/pages/skill-editor-page/editor-tab/skill-misconceptions-editor/skill-misconceptions-editor.directive.ts:
```diff
@@ -161,10 +163,17 @@ angular.module('oppia').directive('skillMisconceptionsEditor', [
-              $scope.misconceptions = $scope.skill.getMisconceptions();
-            });
+            ctrl.parentSubscription.add(
+              SkillEditorStateService.getSkillReinitializedSubject()
```
Reduce the number of like whenever possible. The preferred way is to break after `(`. Please see my suggestions below.

srijanreddy98 replied:
Done!

kevintab95 commented on core/templates/pages/skill-editor-page/editor-tab/skill-description-editor/skill-description-editor.directive.ts:
```diff
@@ -71,10 +74,19 @@ angular.module('oppia').directive('skillDescriptionEditor', [
-              $scope.tmpSkillDescription = $scope.skill.getDescription();
-            });
+            ctrl.parentSubscription.add(
+              SkillEditorStateService.getSkillReinitializedSubject()
```
Reduce lines when possible. Break after `(`.

srijanreddy98 replied:
Done!

kevintab95 commented on core/templates/pages/skill-editor-page/editor-tab/skill-description-editor/skill-description-editor.directive.ts:
```diff
@@ -36,11 +38,12 @@ angular.module('oppia').directive('skillDescriptionEditor', [
-          $scope.MAX_CHARS_IN_SKILL_DESCRIPTION = (
-            MAX_CHARS_IN_SKILL_DESCRIPTION);
+          ctrl.parentSubscription = new Subscription();
+          $scope.MAX_CHARS_IN_SKILL_DESCRIPTION =
```
Wrap inside `(..)` in you have statment spanning multiple lines. Break after `(` if required.

srijanreddy98 replied:
Done!

kevintab95 commented on core/templates/pages/skill-editor-page/editor-tab/skill-concept-card-editor/skill-concept-card-editor.directive.ts:
```diff
@@ -183,9 +186,12 @@ angular.module('oppia').directive('skillConceptCardEditor', [
-              initBindableFieldsDict();
-            });
+            ctrl.parentSubscription.add(
+              SkillEditorStateService.getSkillReinitializedSubject()
```
```suggestion
              SkillEditorStateService.getSkillReinitializedSubject().subscribe(
                () => initBindableFieldsDict()));
```

srijanreddy98 replied:
Done!

kevintab95 reviewed (changes requested):
Hi @srijanreddy98, thanks for the PR. I took a pass and left some nits about code-styling. PTAL!

srijanreddy98 commented:
Hi @kevintab95, thanks for the coding style suggestions, what you suggested is exactly what I had before, but the max line length of 80 forced me to shift the `.subscribe(` to a new line.

kevintab95 commented:
Hi @srijanreddy98, thanks for checking, but for the suggestions I left, I have checked the max char limit and it seems to be under 80 chars. Please double check. I'm fine if that is not the case, but please do comment back on the review comments. Thanks!

srijanreddy98 committed:
Code review

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into migrate-skill-init-event

srijanreddy98 commented:
@kevintab95, Made the requested changes. Could you take a look.

iamprayush commented:
@srijanreddy98 Please note the failing CI tests.

srijanreddy98 commented:
Yeah, those are a flake, I seem have lost the permissions to rerun the test. @kevinlee12, @bansalnitish, @iamprayush. Could one of you just rerun circle ci from failed?

kevinlee12 commented:
did you make sure you are logged into CircleCI using your Github account?

srijanreddy98 commented:
> did you make sure you are logged into CircleCI using your Github account?

Yes, I don't know why but I can click on rerun, but I can't click on any of the options which show up when you click on that button.

srijanreddy98 commented:
@kevinlee12 could you rerun the e2e_topic_and_story_editor

kevinlee12 commented:
> > did you make sure you are logged into CircleCI using your Github account?
> 
> 
> 
> Yes, I don't know why but I can click on rerun, but I can't click on any of the options which show up when you click on that button.

Hmm, okay

srijanreddy98 commented:
> @kevintab95, Made the requested changes. Could you take a look?

@kevintab95 ping!

kevinlee12 commented:
@srijanreddy98 can you merge from develop?

kevinlee12 commented:
your circleci users suite didn't run since you don't have the latest config.

srijanreddy98 commented:
> @srijanreddy98 can you merge from develop?

Ok, I will do it

srijanreddy98 committed:
Merge remote-tracking branch 'upstream/develop' into migrate-skill-init-event

kevintab95 reviewed (approved):
LGTM, thanks @srijanreddy98!

seanlip commented on core/templates/pages/skill-editor-page/editor-tab/skill-concept-card-editor/skill-concept-card-editor.directive.ts:
```diff
@@ -54,9 +56,10 @@ angular.module('oppia').directive('skillConceptCardEditor', [
+        '$scope', '$filter', '$uibModal',
+        function($scope, $filter, $uibModal) {
           var ctrl = this;
+          ctrl.parentSubscription = new Subscription();
```
Can we use a name here that indicates more clearly what type of thing the subscription represents? When I read this I don't really know what "parent" refers to, so it feels a bit like you're naming something "unicodeString" (i.e., the name is correct but not conveying much meaning). It might be better to pick a name that is more related to the type of events that the subscription handles (say).

srijanreddy98 replied:
Maybe `skillChangeSubscription` or I was thinking along the lines of `subscriptionsTracker`?

seanlip replied:
Probably the former, in terms of format. The latter isn't so good because it sounds like a "tracker" object whereas this is clearly a "subscription" object.

But let's dig into the semantic meaning. Why "skill change"? Do initialization/reinitialization events fire only on a change of the skill ID?

srijanreddy98 replied:
Changed this to `directiveSubscriptions`

seanlip commented on core/templates/pages/skill-editor-page/editor-tab/skill-description-editor/skill-description-editor.directive.ts:
```diff
@@ -36,9 +38,10 @@ angular.module('oppia').directive('skillDescriptionEditor', [
+        '$scope',
+        function($scope) {
           var ctrl = this;
+          ctrl.parentSubscription = new Subscription();
```
Ditto -- make this name more descriptive.

seanlip commented on core/templates/pages/skill-editor-page/editor-tab/skill-misconceptions-editor/skill-misconceptions-editor.directive.ts:
```diff
@@ -42,13 +44,13 @@ angular.module('oppia').directive('skillMisconceptionsEditor', [
-        function(
-            $scope, $filter, $uibModal, $rootScope,
-            MisconceptionObjectFactory, EVENT_SKILL_REINITIALIZED,
+        function( $scope, $filter, $uibModal, $rootScope,
```
When stuff in parens spans a newline, break after the opening paren, like it was before.

srijanreddy98 replied:
Got it, will change it.

seanlip replied:
Not changed. Please check "Files Changed" before you request a review, and make sure you have responded to all open comments before doing so, otherwise things like this will delay the review process.

srijanreddy98 replied:
Done!

seanlip commented on core/templates/pages/skill-editor-page/editor-tab/skill-misconceptions-editor/skill-misconceptions-editor.directive.ts:
```diff
@@ -42,13 +44,13 @@ angular.module('oppia').directive('skillMisconceptionsEditor', [
+            MisconceptionObjectFactory,
             MAX_CHARS_IN_MISCONCEPTION_NAME) {
           var ctrl = this;
+          ctrl.parentSubscription = new Subscription();
```
Ditto re naming.

seanlip commented on core/templates/pages/skill-editor-page/editor-tab/skill-rubrics-editor/skill-rubrics-editor.directive.ts:
```diff
@@ -35,22 +37,28 @@ angular.module('oppia').directive('skillRubricsEditor', [
-            RubricObjectFactory, EVENT_SKILL_REINITIALIZED) {
+            RubricObjectFactory) {
           var ctrl = this;
+          ctrl.parentSubscription = new Subscription();
```
Ditto re naming.

seanlip commented on core/templates/pages/skill-editor-page/navbar/skill-editor-navbar.directive.ts:
```diff
@@ -35,18 +35,13 @@ angular.module('oppia').directive('skillEditorNavbar', [
+            $scope, $uibModal, AlertsService, UndoRedoService,
+            SkillEditorStateService, SkillEditorRoutingService) {
           var ctrl = this;
+          ctrl.subscriptions = [];
```
Is this used?

srijanreddy98 replied:
No, It has to be removed, this was the old way of keeping track of subscriptions.

srijanreddy98 replied:
Removed!

seanlip commented on core/templates/pages/skill-editor-page/questions-tab/skill-questions-tab.directive.ts:
```diff
@@ -50,17 +52,16 @@ angular.module('oppia').directive('questionsTab', [
-            EVENT_SKILL_INITIALIZED, EVENT_SKILL_REINITIALIZED,
             NUM_QUESTIONS_PER_PAGE) {
           var ctrl = this;
+          ctrl.parentSubscription = new Subscription();
```
Ditto re naming.

seanlip reviewed (changes requested):
Thanks, took a pass. 

@srijanreddy98 Could you explain how you tested this manually to ensure that all the existing functionality still works? What manual tests did you carry out?

srijanreddy98 commented:
> @srijanreddy98 Could you explain how you tested this manually to ensure that all the existing functionality still works? What manual tests did you carry out?

Sure!

I started by looking where the broadcast was present. It was present in this function `var _setSkill = function(skill) {`  in `skill-editor-state.service`.
This function (`_setSkill`) was only called in one other function (`_updateSkill`)  in the same file.
`_updateSkill` was called in only two functions `loadSkill` and `setSkill`. I checked in the skill module where these functions were called, added console.logs to the `Subject.subscribe()` to check the control reached there.

For example:
`saveSkill` is triggered when we save skill (and there are no unsaved changes but the button is greyed out anyways). I edited the skill in skill editor and saved it and found the expected console.log() outputs.

seanlip commented:
That sounds like a good start. As a sanity check, could you describe the effect of the passed message? What would happen, from a user-visible perspective, if the "initialized"/"reinitialized" message was not received? It is important to check that too, in order to ensure that the message-passing makes sense and is actually needed.

srijanreddy98 commented:
> effect of the passed message

Didn't quite get what you mean?

> What would happen, from a user-visible perspective, if the "initialized"/"reinitialized" message was not received

The only case I can think of such an occurrence happening is when the directive is initialized after the event was broadcasted. In that case, the $onInit should take of proper initialization which is the case in every directive I modified.

>  It is important to check that too, in order to ensure that the message-passing makes sense and is actually needed.

Yes, I checked for that too. I removed two instances of `$scope.$on` from the navbar because it was not required. You can see that in the changes in `skill-editor-navbar.directive.ts`.

seanlip commented:
>> effect of the passed message

> Didn't quite get what you mean?

Presumably the message is passed in order to make something happen. What actually happens and how is this manifested to the user?

>> What would happen, from a user-visible perspective, if the "initialized"/"reinitialized" message was not received

I think you answered the question "Would it be possible for the message to not be received". I'm asking a different question, which is "What is the user-visible effect if the message is not received" -- e.g. does the page fail to load, will the skill be wrong, etc.? This is basically the opposite of the previous question.

srijanreddy98 commented:
> I think you answered the question "Would it be possible for the message to not be received". I'm asking a different question, which is "What is the user-visible effect if the message is not received" -- e.g. does the page fail to load, will the skill be wrong, etc.? This is basically the opposite of the previous question.

Ah! got it. So In a situation when this mechanism doesn't work (which I highly doubt would happen), it would have the same effect if the `$scope.$on` didn't work. It will not get the updated misconceptions in case of `skill-misconceptions-editor.directive.ts` and similarly whatever code was designated to run when the event happened.

seanlip commented:
Yep -- but what happens observably to the user (who is not reading the code, but looking at / trying to interact with the page)? Try to set up that situation and test it.

srijanreddy98 commented:
> Presumably the message is passed in order to make something happen. What actually happens and how is this manifested to the user?

I am assuming here that the "passed message" means the `subject.next()` triggering all the active subscriptions of the subject. So these subjects are used to notify the code of some event happening (user-based event or the time the skill is fetched from server). So it will execute whatever code was kept inside the subscribe function whenever the event happens. So to answer how is this manifested to the user will be different for each case. So in case of lets say `skill-misconceptions-editor.directive.ts` the user will see updated `skill-misconceptions`.

srijanreddy98 commented:
> Yep -- but what happens observably to the user (who is not reading the code, but looking at / trying to interact with the page)? Try to set up that situation and test it.

Sure. I will check and let you know.

srijanreddy98 committed:
Rename to diretiveSubscriptions

srijanreddy98 commented:
@seanlip if the subject mechanism fails for some reason, the user will no longer see updates related to that event. So, in this case, the user will continue seeing old data until the user refreshes the page.

seanlip commented on core/templates/pages/skill-editor-page/editor-tab/skill-concept-card-editor/skill-concept-card-editor.directive.ts:
```diff
@@ -183,9 +186,10 @@ angular.module('oppia').directive('skillConceptCardEditor', [
-            $scope.$on(EVENT_SKILL_REINITIALIZED, function() {
-              initBindableFieldsDict();
-            });
+            ctrl.directiveSubscriptons.add(
```
subscriptions is misspelled. (Make sure to proofread your PRs.)

srijanreddy98 replied:
Changed!

seanlip reviewed (changes requested):
Thanks, LGTM except for a few small inline comments.

Also, just one general question -- is there any issue with combining "skill initialized" and "skill reinitialized" into a single event? What goes wrong if we try to do that?

srijanreddy98 committed:
Code review

srijanreddy98 committed:
Merge initialized and reinitialized subject

srijanreddy98 commented:
> Also, just one general question -- is there any issue with combining "skill initialized" and "skill reinitialized" into a single event? What goes wrong if we try to do that?

@seanlip 
This came across my mind too. This feels like a distinction without a difference (or at least from their usage in the codebase). I merged these two and didn't find any difference from a user perspective.

srijanreddy98 committed:
Code review

srijanreddy98 commented:
@seanlip it is ready for another pass, ptal

seanlip reviewed (approved):
Thanks @srijanreddy98! No further concerns.

seanlip commented:
@bansalnitish, please note, there is one comment for you. Thanks!

srijanreddy98 commented:
@seanlip this can be merged now, right?

seanlip commented:
Yes, I think so. Thanks for the ping @srijanreddy98! Merging.
